### DISCUSSION TITLE
Consumers disconnect after long idle from RabbitMQ queue

### DISCUSSION THREAD

**Discussion by stefanofal-altilia** [pika/pika#1418] [Q&A]
In my company we are currently facing a strange behavour using rabbitMQ. One of our microservices hosted in a Kubernetes Cluster is built in **Python** (the others in **Java** and they don't show this problem) and communicates using Pika implementation for RabbitMQ.

During the week the microservice works just fine, in the weekend it is not used and still up and running. **On monday** trying to publish a message on this microservice queue results in them remaining unacked because **all the consumers have disconnected**. Looking in the code logs, there is absolutely nothing: no error, no exception thrown, nothing at all. The service remains idle without noticing it is disconnected from the queue.

In the first implementation we had a BlockingConnection, but referencing this issue https://github.com/pika/pika/issues/877 and suspecting some kind of idle termination, we moved to **SelectConnection** using ioloops. The heartbeat is configured with standard 60s with broker negotation. I assume it is not a problem in the rabbitMQ broker because every other microservice in Java is working just fine and the connection keeps alive for weeks.

I could not find anything relevant on the logs of rabbitMQ either because of all the other microservices' noise or due to the behavour being silent, maybe. 

About our configuration: 
**Pika** version 1.2.0
**Python** 3.8.3
**RabbitMQ** 3.8.18
**Erlang** 23.3.4.4

The reproducible code can be found here: [github link to repo](https://github.com/stefanofal-altilia/idle-disconnecting-rabbitmq/blob/main/consumer.py)

This is a crosspost of [this stackoverflow question](https://stackoverflow.com/questions/75523362/why-do-consumers-disconnect-from-the-queue-silently-during-long-idles-using-pika?noredirect=1#comment133275317_75523362) following @lukebakken kind suggestion.

Current **logs** for the three RabbitMQ instances running in the Cluster.
[1.txt](https://github.com/pika/pika/files/10814923/1.txt)
[2.txt](https://github.com/pika/pika/files/10814926/2.txt)
[3.txt](https://github.com/pika/pika/files/10814927/3.txt)

**Comment by lukebakken** [comment]
Hello! I _really_ need to know what RabbitMQ logs in your environment. Please attach the logs (or add them to your repo), even if you don't think they contain anything relevant. Thanks.

**Reply by stefanofal-altilia** [reply]
Hello, I edited the original post adding the logs of our three RabbitMQ instances as requested. Thanks

**Reply by lukebakken** [reply]
Thanks. Yep the logs don't show anything relevant to this issue, but you do have connection churn with an application. Usually this means that an application is opening a new connection to publish a single message. This isn't a huge deal if publishing is infrequent, but otherwise your apps should use long-running connections.

So all I need to do to reproduce this is run the `consumer.py` program?

**Reply by stefanofal-altilia** [reply]
Thank you for the feedback about connection churn, I probably found it in our application, currently there is no re-use of connection and channel in a microservice. 
I am not sure how to reproduce the exact issue since the microservice using consumer.py works without any problems during the week but stops in the weekend, I was hoping there is something wrong in the configuration used by the python consumer.

**Reply by lukebakken** [reply]
The consumer program looks fine.

* Does the disconnection happen at the same time, every time?
* Is there a load balancer in your environment?
* Are your Java apps using exactly the same environment as the Python ones?

**Reply by stefanofal-altilia** [reply]
1. I could not pinpoint the exact time during the weekend when the consumers disconnected, but it happens on all the instances across all the clusters. None has a different behavour.
2. Not yet, we are considering to try to mitigate this problem scaling microservices using pending messages on rabbitMQ queues, but it would be much more preferrable to look into a working solution client-side.
3. Yes, they all share the same environments (cluster and namespace, connecting to the same RabbitMQ broker) and this behavour happens across all Python microservices but not with Java ones.

**Reply by lukebakken** [reply]
Would it be possible to enable `INFO` level debugging in one of your Python consumers so we can see what the heck happens during the weekend?

https://github.com/pika/pika/blob/main/examples/asynchronous_consumer_example.py#L10-L12

https://github.com/pika/pika/blob/main/examples/asynchronous_consumer_example.py#L433

**Reply by stefanofal-altilia** [reply]
Yes, I updated the github consumer.py to strictly follow the changed I did in our DEV environment using the logging level you pointed. 
After I updated our internal project I saw that the Kafka broker now is logged too, I hope it is not a problem. On Monday I will provide the complete logs from our DEV cluster and the filtered ones grepping Pika. 
Thank you again for your help.

**Reply by stefanofal-altilia** [reply]
After friday deploy enabling the logging, I noticed the microservice running on localhost had **different logs** than the one I pushed in our cluster. Particularly, I saw that were missing the "**send_heartbeat**" calls by Pika, so I further investigated both my docker configuration (where I found **heartbeat setted to 60s**) and our cluster instance which I discovered had it set to 0 (meaning no heartbeat, if I am not mistaken).
Following this, I decided to push the code with logging on two different Kubernetes clusters, let's call them A and B. 
On A, I left rabbitMQ broker with heartbeat at 0, on B I set it to 60s. 
On monday, I found that consumers in **A** were **silently dead**, as it always happened 'til now, but on **B** they were **still connected** and the microservice was up and running.
I post this for two reasons: 
1. to provide a suggestion to anyone who might happen to have the same problem, check the heartbeat effictively set by your broker after enabling the logs as @lukebakken suggested;
2. to post the logs as we agreed.

I intend to push during this week the new code to almost all our clusters and see during the next weekend if this hideous bug is solved.

[logs_cluster_A.txt](https://github.com/pika/pika/files/10840451/logs_cluster_A.txt)
[logs_cluster_B.txt](https://github.com/pika/pika/files/10840452/logs_cluster_B.txt)

**Reply by lukebakken** [reply]
Thanks for the update! Having "silently dead" consumers isn't good, I'll see if I can reproduce this and fix that.

Would you mind providing your *full* "good" and "not good" RabbitMQ configurations just to be sure I'm using exactly the same settings?

**Reply by stefanofal-altilia** [reply]
Yes, I can provide the configurations. Can you be more specific about what you would like to get, and where to get it? Thanks.

**Reply by lukebakken** [reply]
`rabbitmq.conf`, `advanced.config` if you are using that. They should be in `/etc/rabbitmq` but you're using Kubernetes so who knows? Providing the RabbitMQ part of that config may be enough.

**Reply by stefanofal-altilia** [reply]
Here the configurations, there is no advanced.config that I could find. I removed only the password for safety concerns from configuration_A.
[configurations.zip](https://github.com/pika/pika/files/10841350/configurations.zip)

**Reply by stefanofal-altilia** [reply]
Hello @lukebakken, sorry to bother you again, but I am running in an issue. One of the microservices in Python crashed, logs are enabled so I could see what is happening (logs attached filtering for pika). 
```
Connection._on_connection_close_from_broker of <SelectConnection OPEN transport=<pika.adapters.utils.io_services_utils._AsyncPlaintextTransport object at 0x7f9f7ced20d0> params=<ConnectionParameters host=rabbitmq.rabbitmq-system.svc.cluster.local port=5672 virtual_host=/ ssl=False>>> for "0:Connection.Close"
DEBUG      2023-03-01 12:32:36,322 pika.connection                _on_connection_close_from_broker     1777: _on_connection_close_from_broker: frame=<METHOD(['channel_number=0', 'frame_type=1', 'method=<Connection.Close([\'class_id=0\', \'method_id=0\', \'reply_code=320\', "reply_text=CONNECTION_FORCED - broker forced connection closure with reason \'shutdown\'"])>'])>
```
I think this is the reason why the consumer is disconnected, but I am questioning why no exception was raised in the code. Attached there is the filtered log and the complete one from our A cluster.

If you prefer, I could open another issue here on github. Thanks.
[filtered_logs.txt](https://github.com/pika/pika/files/10863541/stop_timeout.txt)
[complete_logs.txt](https://github.com/pika/pika/files/10863552/1.txt)

**Reply by lukebakken** [reply]
What was logged by RabbitMQ around the same point in time?

**Reply by stefanofal-altilia** [reply]
Unfortunetly, our logs don't go so far to that time, but I found something interesting. In our configuration, we have three pods running rabbitmq, and coincidentally one of them restarted exactly when this error was thrown. 
This explains why the connection failed and the broker closed it, but I would like to understand if it is usually silent in Pika or an Exception is thrown.

**Reply by lukebakken** [reply]
The Pika message you provided is exactly what I would expect to see when the connected broker instance is restarted.

**Reply by urvisism** [reply]
I have been also facing this issue in RabbitMQ.
I tried your solution to set Heartbeat value in rabbitmq.conf file.
But, after that It doesn't take RABBITMQ_USERNAME and RABBITMQ_PASSWORD in consideration, which I have written in docker-compose.yaml file and I want to use above credentials from .env file, which I can't do after using .conf file. The docker-compose file is given bellow.
So, Is it compulsory to set Heartbeat value on both Docker and Consumer side ? or If I set it on consumer side only, then also It'll work ?
```
networks:
  rabbitmq:
    driver: bridge
    ipam:
      driver: default
      config:
        - subnet: 172.16.255.0/24
          gateway: 172.16.255.1

services:

  rabbitmq:
    image: bitnami/rabbitmq:3.12.12
    container_name: rabbitmq
    hostname: rabbitmq
    ports:
      - "5672:5672"
      - "15672:15672"
    environment:
      - RABBITMQ_USERNAME=${RABBITMQ_USERNAME:-user}
      - RABBITMQ_PASSWORD=${RABBITMQ_PASSWORD:-root}
      - RABBITMQ_MANAGEMENT_ALLOW_WEB_ACCESS=true
      - RABBITMQ_PLUGINS=rabbitmq_management,rabbitmq_shovel,rabbitmq_shovel_management,rabbitmq_delayed_message_exchange
      - RABBITMQ_COMMUNITY_PLUGINS=https://github.com/rabbitmq/rabbitmq-delayed-message-exchange/releases/download/v3.12.0/rabbitmq_delayed_message_exchange-3.12.0.ez
    volumes:
      - rabbitmq_data:/bitnami/rabbitmq/mnesia
      - ./config/rabbitmq/rabbitmq.conf:/opt/bitnami/rabbitmq/etc/rabbitmq/rabbitmq.conf:rw
    networks:
      - rabbitmq

volumes:
  rabbitmq_data:
```

**Reply by stefanofalangone** [reply]
@urvisism here OP! From my experience with this issue and what I managed to test, you have to put the heartbeat also on the Docker file. I had this exact problem setting up it only on Producer side and I had the behavior I explained in the original post.
I am not specialized in devOps, but it is definetly possible to setup both credentials and heartbeat from a external file, at least in a kubernetes yaml. 
We do that at my company, even though I am not in charge of it.
Hope this helps a bit!
